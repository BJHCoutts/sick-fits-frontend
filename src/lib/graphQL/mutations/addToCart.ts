import { gql } from "@apollo/client"

export const ADD_TO_CART_MUTATION = gql`
	mutation ADD_TO_CART($productId: ID!) {
		addToCart(productId: $productId) {
			id
		# quantity
		# product {
		# 	id
		# 	name
		# }
		# user {
		# 	id
		# 	name
		# 	email
			# }
		}
	}
`