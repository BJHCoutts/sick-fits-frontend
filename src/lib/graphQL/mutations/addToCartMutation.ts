import { gql } from "@apollo/client"

export const ADD_TO_CART_MUTATION = gql`
	mutation ADD_TO_CART_MUTATION($productId: ID!) {
		addToCart(productId: $productId) {
			id
		}
	}
`