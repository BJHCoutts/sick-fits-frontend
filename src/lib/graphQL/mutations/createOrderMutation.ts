import { gql } from "@apollo/client"

export const CREATE_ORDER_MUTATION = gql`
	mutation CREATE_ORDER_MUTATION($token:String!) {
  checkout (token:$token) {
    id
    charge
    total
    items{
      id
      name
    }
  }
}
`