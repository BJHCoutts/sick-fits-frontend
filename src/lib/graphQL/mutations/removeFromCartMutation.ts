import { gql } from "@apollo/client"

export const REMOVE_FROM_CART_MUTATION = gql`
	mutation REMOVE_FROM_CART_MUTATION ($id:ID!){
		deleteCartItem (id:$id) {
			id
			quantity
			product {
				name
			}
		}
	}
`