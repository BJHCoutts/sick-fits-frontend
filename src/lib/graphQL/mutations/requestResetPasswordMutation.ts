import { gql } from "@apollo/client"

export const REQUEST_RESET_PASSWORD_MUTATION = gql`
	mutation REQUEST_RESET_PASSWORD_MUTATION($email: String!) {
		sendUserPasswordResetLink (email:$email) {
      code
      message
    }
	}
`