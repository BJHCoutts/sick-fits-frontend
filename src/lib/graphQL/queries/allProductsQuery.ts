import { gql } from "@apollo/client"

export const ALL_PRODUCTS_QUERY = gql`
query ALL_PRODUCTS_QUERY{
  allProducts {
    id
    name
    description
    photo {
      id
      image {
        id
        publicUrlTransformed
      }
      altText
    }
    status
  }
}
`