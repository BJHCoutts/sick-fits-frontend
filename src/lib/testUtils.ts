import casual from 'casual'
import { PAGINATION_QUERY } from './graphQL/queries/paginationQuery'
import { TProduct } from './types/TProduct'
import { TCartItem } from './types/TCartItem'
import { TUser } from './types/TUser'

// seed it so we get consistent results
casual.seed(777)

export const fakeItem = (): TProduct => ({
	__typename: 'Item',
	id: 'abc123',
	price: 5000,
	user: null,
	photo: {
		id: 'abc123',
		altText: 'dogs are best',
		image: {
			id: 'abc123',
			publicUrlTransformed: 'dog.jpg',
		},
	},
	name: 'dogs are best',
	description: 'dogs',
	status: 'AVAILABLE',
})

export const fakeUser = (overrides: {}): TUser => ({
	__typename: 'User',
	id: '4234',
	name: casual.name,
	email: casual.email,
	permissions: ['ADMIN'],
	orders: [],
	cart: [],
	...overrides,
})

// const fakeOrderItem = () => ({
//   __typename: 'OrderItem',
//   id: casual.uuid,
//   image: {
//     image: `${casual.word}.jpg`,
//   },
//   name: casual.words(),
//   price: 4234,
//   quantity: 1,
//   description: casual.words(),
// });

// const fakeOrder = () => ({
//   __typename: 'Order',
//   id: 'ord123',
//   charge: 'ch_123',
//   total: 40000,
//   items: [fakeOrderItem(), fakeOrderItem()],
//   createdAt: '2022-12-11T20:16:13.797Z',
//   user: fakeUser(),
// });

export const fakeCartItem = (overrides: {}): TCartItem => ({
	__typename: 'CartItem',
	id: 'omg123',
	quantity: 4,
	product: fakeItem(),
	user: fakeUser({}),
	...overrides,
})

// // Fake LocalStorage
// class LocalStorageMock {
//   constructor() {
//     this.store = {};
//   }

//   clear() {
//     this.store = {};
//   }

//   getItem(key) {
//     return this.store[key] || null;
//   }

//   setItem(key, value) {
//     this.store[key] = value.toString();
//   }

//   removeItem(key) {
//     delete this.store[key];
//   }
// }

export function makePaginationMocksFor(length: number) {
	return [
		{
			request: { query: PAGINATION_QUERY },
			result: {
				data: {
					_allProductsMeta: {
						count: length,
					},
					itemsConnection: {
						__typename: 'aggregate',
						aggregate: {
							count: length,
							__typename: 'count',
						},
					},
				},
			},
		},
	]
}

// export
// makePaginationMocksFor,
// LocalStorageMock,
// fakeItem,
// fakeUser,
// fakeCartItem,
// fakeOrder,
// fakeOrderItem,
