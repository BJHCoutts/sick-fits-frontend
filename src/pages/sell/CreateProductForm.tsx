import { useMutation } from '@apollo/client'
import SForm from '../../components/styles/SForm'
import useForm from '../../lib/functions/useForm'
import { CREATE_PRODUCT_MUTATION } from '../../lib/graphQL/mutations/createProductMutation'
import DisplayError from '../../components/ErrorMessage'
import { ALL_PRODUCTS_QUERY } from '../../lib/graphQL/queries/allProductsQuery'
import { useRouter } from 'next/router'

export default function CreateProductForm() {
	const router = useRouter()
	const { inputs, handleChange, clearForm } = useForm({
		image: '',
		name: '',
		price: '',
		description: '',
	})

	const [createProduct, { loading, error, data }] = useMutation(
		CREATE_PRODUCT_MUTATION,
		{
			variables: inputs,
			refetchQueries: [{ query: ALL_PRODUCTS_QUERY }],
		}
	)

	async function handleSubmit(e: React.SyntheticEvent) {
		e.preventDefault()
		console.log('PRESSED')
		const res = await createProduct()
		console.log(inputs.image)
		clearForm()
		router.push({
			pathname: `/product/${res.data.createProduct.id}`,
		})
	}

	return (
		<>
			<SForm>
				<DisplayError error={error} />
				<fieldset disabled={loading} aria-busy={loading}>
					<label htmlFor="image">
						Image
						<input
							type="file"
							id="image"
							name="image"
							onChange={handleChange}
							required
						/>
					</label>
					<label htmlFor="name">
						Name
						<input
							type="text"
							id="name"
							name="name"
							placeholder="name"
							value={inputs.name}
							onChange={handleChange}
							required
						/>
					</label>
					<label htmlFor="price">
						Price
						<input
							type="number"
							id="price"
							name="price"
							placeholder="0"
							value={inputs.price}
							onChange={handleChange}
							required
						/>
					</label>
					<label htmlFor="description">
						Description
						<textarea
							id="description"
							name="description"
							placeholder="description"
							value={inputs.description}
							onChange={handleChange}
						/>
					</label>

					<div className="button-container">
						<button type="button" onClick={clearForm}>
							Clear Form
						</button>
						<button type="button" onClick={handleSubmit}>
							+ Add Product
						</button>
					</div>
				</fieldset>
			</SForm>
		</>
	)
}
