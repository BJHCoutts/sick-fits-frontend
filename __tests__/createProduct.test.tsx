import { render, screen, waitFor } from '@testing-library/react'
import { fakeItem } from '../src/lib/testUtils'
import CreateProductForm from '../src/pages/sell/CreateProductForm'
import { MockedProvider } from '@apollo/client/testing'
import userEvent from '@testing-library/user-event'
import { ALL_PRODUCTS_QUERY } from '../src/lib/graphQL/queries/allProductsQuery'
import wait from 'waait'
import { CREATE_PRODUCT_MUTATION } from '../src/lib/graphQL/mutations/createProductMutation'

const item = fakeItem()

jest.mock('next/router', () => ({
	useRouter: jest.fn(),
}))

const testImageFile = new File(['hello'], 'hello.png', {
	type: 'image/png',
})

const useRouter = jest.spyOn(require('next/router'), 'useRouter')

describe('<CreateProduct/>', () => {
	it('renders and matches snapshot', () => {
		const { container, debug } = render(
			<MockedProvider>
				<CreateProductForm />
			</MockedProvider>
		)
		expect(container).toMatchSnapshot()
	})

	it('handles the updating', async () => {
		const { container, debug } = render(
			<MockedProvider>
				<CreateProductForm />
			</MockedProvider>
		)

		await userEvent.type(screen.getByLabelText('Name'), item.name)
		await userEvent.type(screen.getByLabelText('Price'), item.price.toString())
		await userEvent.type(
			screen.getByLabelText(/description/i),
			item.description
		)

		expect(screen.getByDisplayValue(item.name)).toBeInTheDocument()
		expect(screen.getByDisplayValue(item.price.toString())).toBeInTheDocument()
	})

	it('creates an item when the form is submitted', async () => {
		const mocks = [
			{
				request: {
					query: CREATE_PRODUCT_MUTATION,
					variables: {
						name: item.name,
						description: item.description,
						image: testImageFile,
						price: item.price,
					},
				},
				result: {
					data: {
						createProduct: {
							...item,
							id: 'abc123',
							__typename: 'Item',
						},
					},
				},
			},
			{
				request: {
					query: ALL_PRODUCTS_QUERY,
					variables: { skip: 0, first: 2 },
				},
				result: {
					data: {
						allProducts: [item],
					},
				},
			},
		]

		const { container, debug } = render(
			<MockedProvider>
				<CreateProductForm />
			</MockedProvider>
		)

		const push = jest.fn()

		useRouter.mockImplementation(() => ({
			push,
		}))

		await userEvent.type(screen.getByLabelText('Name'), item.name)
		await userEvent.type(screen.getByLabelText('Price'), item.price.toString())
		await userEvent.type(
			screen.getByLabelText(/description/i),
			item.description
		)

		const fileInput = screen.getByLabelText(/image/i) as HTMLInputElement

		if (fileInput.files) {
			expect(fileInput.files.length).toBe(0)
			await userEvent.upload(fileInput, testImageFile)
			expect(fileInput.files.length).toBe(1)
		}

		await userEvent.click(screen.getByText(/Add Product/i))

		// await waitFor(() => wait(5))

		// debug()

		// expect(push).toHaveBeenCalled()
	})
})
