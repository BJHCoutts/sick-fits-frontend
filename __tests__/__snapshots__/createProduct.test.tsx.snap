// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateProduct/> renders and matches snapshot 1`] = `
<div>
  <form
    class="sc-bdfCDU cdIIGl"
  >
    <fieldset
      aria-busy="false"
    >
      <label
        for="image"
      >
        Image
        <input
          id="image"
          name="image"
          required=""
          type="file"
        />
      </label>
      <label
        for="name"
      >
        Name
        <input
          id="name"
          name="name"
          placeholder="name"
          required=""
          type="text"
          value=""
        />
      </label>
      <label
        for="price"
      >
        Price
        <input
          id="price"
          name="price"
          placeholder="0"
          required=""
          type="number"
          value=""
        />
      </label>
      <label
        for="description"
      >
        Description
        <textarea
          id="description"
          name="description"
          placeholder="description"
        />
      </label>
      <div
        class="button-container"
      >
        <button
          type="button"
        >
          Clear Form
        </button>
        <button
          type="button"
        >
          + Add Product
        </button>
      </div>
    </fieldset>
  </form>
</div>
`;
